src/k2/tso/client/tso_clientlib.cpp

```diff
@@ -497,33 +492,35 @@ seastar::future<TimestampBatch> TSO_ClientLib::GetTimestampBatch(uint16_t batchS
 
             K2ASSERT(log::tsoclient, !_curTSOServiceNodes.empty(), "we should have workers");
             // pick next worker (effecitvely random one, as _curTSOServiceNodes is shuffled already when it is populated)
-            int randNode = (_curWorkerIdx++) %  _curTSOServiceNodes.size();
-
-            auto myRemote = _curTSOServiceNodes[randNode];
-            std::unique_ptr<Payload> payload = myRemote.newPayload();
-            payload->write(batchSize);
+            if (retriesLeft != 2) {_curWorkerIdx++;}  // if this is not first try, it means we had error and are retrying, thus change to a new service node. 
+            int randNode = _curWorkerIdx %  _curTSOServiceNodes.size();
+            auto& myRemote = _curTSOServiceNodes[randNode];
 
-            (void) retriesLeft;
-            // K2LOG_I(log::tsoclient, "Requesting timestampBatch with retriesLeft=" << retriesLeft << ", and timeout=" << k2::usec(timeout).count()
-            //        << "us, with worker " << randNode);
+            GetTimeStampBatchRequest request{.batchSizeRequested = batchSize};
+            K2LOG_D(log::tsoclient, "Requesting timestampBatch of batchsize:{} with retriesLeft:{} and timeout:{} to node:{}", batchSize, retriesLeft, timeout, randNode);
 
-            return k2::RPC().sendRequest(dto::Verbs::GET_TSO_TIMESTAMP_BATCH, std::move(payload), myRemote, timeout)
-            .then([this, &batch](std::unique_ptr<k2::Payload> replyPayload) mutable {
+            return k2::RPC().callRPC<dto::GetTimeStampBatchRequest, dto::GetTimeStampBatchResponse>(dto::Verbs::GET_TSO_TIMESTAMP_BATCH, request, *myRemote, timeout)
+            .then([this, &batch](auto&& response) {
                 if (_stopped)
                 {
                     K2LOG_I(log::tsoclient, "Stopping retry since we were stopped");
                     return seastar::make_exception_future<>(TSOClientLibShutdownException());
                 }
 
-                if (!replyPayload || replyPayload->getSize() == 0)
+                auto& [status, r] = response;
+                if (!status.is2xxOK())
+                {
+                    K2LOG_E(log::tsoclient, "Error during get timestampBatch, code:{}, error message:{}", status.code, status.message);
+                    // currently, we should only have 5xx retryable error
+                    K2ASSERT(log::tsoclient, status.is5xxRetryable(), "GetTimeStampBatch error should be 5xxRetryable.");

```

Comment:
you're throwing on the next line so this assertion I think won't be any different.  Maybe just remove the assertion?

Reply:
The assertion here is just to make sure that future change (adding other error status) is handled. So I will keep it.